src/txn/txn_recover.c

```diff
@@ -633,6 +622,51 @@ __wt_txn_recover(WT_SESSION_IMPL *session)
         WT_ERR(ret);
     }
 
+    /*
+     * We should check whether the history store file exists in the metadata or not. or not. If it
+     * does not, then we should not apply rollback to stable to each table. This might happen if
+     * we're upgrading from an older version. If it does exist in the metadata we should check that
+     * it exists on disk to confirm that it wasn't deleted between runs.
+     *
+     * This needs to happen after we recovery the metadata from the logs, otherwise we can end up in
+     * the situation where the metadata does actually contain the history store but we didn't yet
+     * recover the metadata.
+     */
+    metac->set_key(metac, WT_HS_URI);
+    WT_ERR_NOTFOUND_OK(metac->search(metac), true);
+    if (ret == WT_NOTFOUND) {
+        hs_exists = false;
+    } else {
+        /* Given the history store exists in the metadata validate whether it exists on disk. */
+        WT_ERR(__wt_fs_exist(session, WT_HS_FILE, &hs_exists));
```

Comment:
All of this code movement and where we perform this is almost certainly also needed by the branch I created for WT-5814. I will rebase that branch so it is a sub-branch of this rather than develop.